File: Data Structures & Algorithms/minimum-interval-including-query/submission-0.cpp
class Solution {
public:
    vector<int> minInterval(vector<vector<int>>& intervals, vector<int>& queries) {
        
        // Sort intervals by left endpoint
        sort(intervals.begin(), intervals.end());
        
        int n = intervals.size();
        int q = queries.size();
        
        // Store queries with original index
        vector<pair<int,int>> qIdx;
        for (int i = 0; i < q; i++) {
            qIdx.push_back({queries[i], i});
        }
        
        // Sort queries
        sort(qIdx.begin(), qIdx.end());
        
        // Min heap: {length, right}
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<>> pq;
        
        vector<int> ans(q, -1);
        
        int i = 0; // pointer for intervals
        
        for (auto &it : qIdx) {
            int query = it.first;
            int idx = it.second;
            
            // Add intervals whose left <= query
            while (i < n && intervals[i][0] <= query) {
                int left = intervals[i][0];
                int right = intervals[i][1];
                int len = right - left + 1;
                
                pq.push({len, right});
                i++;
            }
            
            // Remove intervals that can't cover query
            while (!pq.empty() && pq.top().second < query) {
                pq.pop();
            }
            
            // If heap not empty, top has smallest interval
            if (!pq.empty()) {
                ans[idx] = pq.top().first;
            }
        }
        
        return ans;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        ans = 0
        left = 0
        right = len(height)-1

        maxl = height[left]
        maxr = height[right]

        while left < right:
            if maxl < maxr:
                left +=1
                maxl = max(maxl, height[left])
                ans+= maxl - height[left]

            else:
                right -=1
                maxr = max(maxr, height[right])
                ans+= maxr - height[right]

        return ans        